src/main/java/com/scalar/cassy/transferer/RemoteFileSystemFileUploader.java

```diff
@@ -0,0 +1,130 @@
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.apache.commons.lang3.time.DurationFormatUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RemoteFileSystemFileUploader extends RemoteFileSystemTransfererBase
+    implements FileUploader {
+  private static final Logger logger = LoggerFactory.getLogger(RemoteFileSystemFileUploader.class);
+  private static final int NB_PARALLEL_UPLOAD_THREADS = 1;
+
+  @Inject
+  public RemoteFileSystemFileUploader(RemoteFileSystemConnection hostConnection) {
+    super(hostConnection);
+  }
+
+  @Override
+  public Future<Void> upload(Path file, String key, String storageBaseUri) {
+    String remoteFileSystemStoragePath = URI.create(storageBaseUri).getPath();
+    Path targetFilePath =
+        hostConnection.getHostControlSystem().getPath(remoteFileSystemStoragePath, key);
+    File targetFile = new File(targetFilePath.toString());
+    if (targetFile.exists() && targetFile.length() == file.toFile().length()) {
+      logger.info(file + " has been already uploaded.");
+      return CompletableFuture.completedFuture(null);
+    }
+
+    logger.info("Uploading " + file);
+    try {
+      createDirectories(targetFilePath.getParent());
+      copyFile(file, targetFilePath);
+      return CompletableFuture.completedFuture(null);
+    } catch (IOException e) {
+      throw new FileTransferException(e);
+    }
+  }
+
+  @Override
+  public void upload(List<Path> files, BackupConfig config) {
+    String remoteFileSystemStoragePath = URI.create(config.getStoreBaseUri()).getPath();
+    Path dataDir = Paths.get(config.getDataDir());
+    ExecutorService executorService = Executors.newFixedThreadPool(NB_PARALLEL_UPLOAD_THREADS);
+    ExecutorCompletionService<Void> executorCompletionService =
+        new ExecutorCompletionService<>(executorService);
+    logger.info(String.format("Uploading keyspace \"%s\"", config.getKeyspace()));
+    long uploadStart = System.currentTimeMillis();
+
+    // Start uploading files
+    for (Path filePath : files) {
+      Path targetFilePath =
+          computeTargetFilePath(config, remoteFileSystemStoragePath, dataDir, filePath);
+      try {
+        this.createDirectories(targetFilePath.getParent());
+      } catch (IOException e) {
+        throw new FileTransferException(e);
+      }
+      submitFileUploadTask(executorCompletionService, filePath, targetFilePath);
+    }
+
+    // Wait for uploads completion
+    int completedUploads = 0;
+    while (completedUploads < files.size()) {
+      try {
+        executorCompletionService.take().get();
+        completedUploads++;
+      } catch (Exception e) {
+        executorService.shutdownNow();
+        throw new FileTransferException(e);
+      }
```

Comment:
If one of the upload task ends up throwing an exception, all non finished uploads tasks will be stopped before re-throwing the exception